res/res_pjsip_rfc3326.c

```diff
@@ -61,14 +68,31 @@ static void rfc3326_use_reason_header(struct ast_sip_session *session, struct pj
 			if (sscanf(cause, "cause=%30d", &code) != 1) {
 				continue;
 			}
-		} else {
-			continue;
-		}
-		if (cause_q850) {
-			ast_channel_hangupcause_set(session->channel, code & 0x7f);
-			break;
-		} else if (cause_sip) {
-			ast_channel_hangupcause_set(session->channel, ast_sip_hangup_sip2cause(code));
+
+			if (cause_q850) {
+				hangupcause = code & 0x7f;
+				ast_channel_hangupcause_set(session->channel, hangupcause);
+			} else if (cause_sip) {
+				hangupcause = ast_sip_hangup_sip2cause(code);
+				ast_channel_hangupcause_set(session->channel, hangupcause);
+			} else if (cause_preemption) {
+				hangupcause = AST_CAUSE_PRE_EMPTED;
+				ast_channel_hangupcause_set(session->channel, hangupcause);
+			}
+
+			data_size = sizeof(*cause_code) + pj_strlen(&header->hvalue) + 1;
+			cause_code = ast_alloca(data_size);
+			memset(cause_code, 0, data_size);
+			ast_copy_string(cause_code->chan_name, ast_channel_name(session->channel), AST_CHANNEL_NAME);
+			cause_code->tech2_offset = 1;
+			cause_code->ast_cause = hangupcause;
+			ast_copy_pj_str(cause_code->code + 1, &header->hvalue, pj_strlen(&header->hvalue) + 1);
+			peer = ast_channel_bridge_peer(session->channel);
```

Comment:
It is hangup reason and it is useful for the peer. It is similar like in chan_pjsip_incoming_response_update_cause, where response data are stored to both channels and available via HANGUPCAUSE(chan,tech)
First time I tried to send data to peer thru bridge using ast_queue_control_data (like in chan_pjsip_incoming_response_update_cause), but I ended up with a race between hangup and data transport when processed message was BYE. Maybe I can directly call ast_channel_hangupcause_hash_set instead of ast_indicate_data?

Reply:
update_case does not reach across the bridge and directly do things, all of this stuff is supposed to be queued on the originating channel, go through a bridge or whatever it may be, and then get applied to the other channel - if that's not working then that should be examined, determined, and resolved

Reply:
Thank you, I spent some time trying to fix the race, but I did not find out how. Will see if I will get green light to spent more time on it